Feature/2 initial version subbranch (#5)
exasol/row-level-security: doc/design.md

Needs: impl, utest, itest

> ## Query Rewriter Determines User Roles

Reviewer: ```suggestion
## `QueryRewriter` Determines User Roles
```

Please repeat for all other occurrences.

Author: fixed

## Query Rewriter Identifies Protected Tables
`dsn~query-rewriter-identifies-protected-tables~1`

The Query Rewriter identifies a table as protected with row-level security, if that table has a column named `exa_row_roles` or a column named  `exa_row_tenants` or both.

Covers:

* `req~rows-users-are-allowed-to-read~1`
* `req~tables-with-tenants-restrictions~1`

Needs: impl, utest, itest

## Query Rewriter Identifies Unprotected Tables
`dsn~query-rewriter-identifies-unprotected-tables~1`

The Query Rewriter identifies a table as unprotected, if that table does not have a column named `exa_row_roles` or `exa_row_tenants`.
The Query Rewriter does not modify an unprotected table.

Covers:

* `req~unprotected-tables~1`

Needs: impl, utest, itest

> ## Query Rewriter Chooses a Way to Treat Protected Tables

Reviewer: Follow-up issue. See comment in system requirements.

Author: fixed

Needs: impl, utest, itest

## `Query Rewriter` Determines User Roles

## `Query Rewriter` Identifies Protected Tables
`dsn~query-rewriter-identifies-protected-tables~1`

The `Query Rewriter` identifies a table as protected with row-level security, if that table has a column named `exa_row_roles` or a column named  `exa_row_tenants` or both.

Covers:

* `req~rows-users-are-allowed-to-read~1`
* `req~tables-with-tenants-restrictions~1`

Needs: impl, utest, itest

## `Query Rewriter` Identifies Unprotected Tables
`dsn~query-rewriter-identifies-unprotected-tables~1`

The `Query Rewriter` identifies a table as unprotected, if that table does not have a column named `exa_row_roles` or `exa_row_tenants`.
The `Query Rewriter` does not modify an unprotected table.

Covers:

* `req~unprotected-tables~1`

Needs: impl, utest, itest

> ## `Query Rewriter Treats Protected Tables with Both Roles and Tenants Restrictions

Reviewer: ```suggestion
## `QueryRewriter` Treats Protected Tables with Both Roles and Tenants Restrictions
```

Author: fixed

`query-rewriter-treats-protected-tables-with-both-roles-and-tenants-restrictions~1`

> If a table contains both `exa_row_roles` and `exa_row_tenants`columns, then the `Query Rewriter` applies both security conceptions. 

Reviewer: ```suggestion
If a table contains both `exa_row_roles` and `exa_row_tenants`columns, then the `Query Rewriter` applies both security schemes. 
```

Author: fixed

That means a user has to be marked as a tenant and have the right role in due to see a row's content.

Covers:

* `req~tables-with-both-roles-and-tenants-restrictions~1`

Needs: impl, utest, itest

> ## `Query Rewriter` Replaces Tables

Reviewer: ```suggestion
## `QueryRewriter` Replaces Tables
```

Author: fixed

`dsn~query-rewriter-replaces-tables~1`

> If a table is protected with row level security, the `Query Rewriter` replaces this table with sub-select that only yields columns the user is allowed to read.

Reviewer: ```suggestion
If a table is protected with row level security, the `QueryRewriter` replaces this table with sub-select that only yields columns the user is allowed to read.
```

Author: fixed

Covers:

* `req~rows-users-are-allowed-to-read~1`

Needs: impl, utest, itest

> ## `Query Rewriter` Adds Row Filter For Roles

Reviewer: ```suggestion
## `Query Rewriter` Adds Row Filter for Roles
```

Author: fixed

`dsn~query-rewriter-adds-row-filter-for-roles~1`

> The `Query Rewriter` adds a row filter to the injected sub-query that uses bitwise-`AND` against the users role mask and checks whether the result is not zero.

Reviewer: ```suggestion
The `QueryRewriter` adds a row filter to the injected sub-query that uses bitwise-`AND` against the users role mask and checks whether the result is not zero.
```

Author: fixed
